CIPs/cip-draft-bls-publickey.md

```diff
@@ -0,0 +1,68 @@
+---
+cip: 
+title: Precompile for getting BLS public key for a given validator in a historical block
+author: Sami Mäkelä (@mrsmkl)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category (*only required for Standards Track): <Ring 0, 1, 2, 3>
+created: 2020-11-17
+license: Apache 2.0
+---
+
+
+## Simple Summary
+
+Adding a precompile that returns the BLS public key of a given validator in a given block.
+
+## Abstract
+
+Returns the BLS public key of a validator in the same way as the existing precompile for getting the validator address.
+
+## Motivation
+
+Currently it's not possible to get the validator BLS public key used for consensus from the chain. It's only possible to get the current BLS key that would be active in the next epoch. However the keys are needed to implement slashing in the case of bad epoch snark data (see CIP-20).
+
+## Specification
+
+The number of precompile is 0x... The input is interpreted in the same way as precompile for getting the validator address: it is 64 bytes, first 32 bytes are an integer representing the index of validator to get, and the second 32 bytes are an integer representing the block to access. These input values have the same conditions as the precompile for getting validator address: index must be smaller than the number of validator for the given block, and block number must be smaller than current block number.
+
+Returns the BLS public key in uncompressed format as 192 byte string. In the case of Celo, the BLS public key is a point in the BLS12-377 elliptic curve defined over Fp2 (G2). A value of Fp2 is represented as two numbers smaller than the base field modulus. These numbers are encoded as 48 byte big endian strings (function `BE`). So if we have a point (a+bi,c+di), it is serialized as `BE(a) || BE(b) || BE(c) || BE(d)`.
```

Comment:
any thought given to returning 256 bytes (4 64-byte numbers) to maintain evm word alignment? with 48-byte numbers, the caller will always have to read the middle of some memory words and mask appropriately

Reply:
It's probably best to return what's most efficient to handle in the EVM, I just described what's happening in the rust code for now.

Reply:
using the 64-byte format instead of the 48-byte format would also match the EIP2537/2539 format and allow you to directly pass the returned block to the precompile (if you manage memory carefully). I think I'd strongly recommend switching to 64-byte